File: android/src/main/java/com/combateafraude/face_authenticator/FaceAuthenticatorPlugin.java
package com.combateafraude.face_authenticator;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;

import androidx.annotation.Nullable;
import androidx.annotation.NonNull;

import java.util.HashMap;

import input.CafStage;
import input.FaceAuthenticator;
import input.VerifyAuthenticationListener;
import input.iproov.Filter;
import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.embedding.engine.plugins.activity.ActivityAware;
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry;
import output.FaceAuthenticatorResult;

@SuppressWarnings("unchecked")
public class FaceAuthenticatorPlugin
        implements FlutterPlugin, MethodCallHandler, ActivityAware, PluginRegistry.ActivityResultListener {

    private MethodChannel channel;
    private Result result;
    private Activity activity;
    private ActivityPluginBinding activityBinding;
    private Context context;

    @Override
    public synchronized void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
        if (call.method.equals("start")) {
            this.result = result;
            start(call);
        } else {
            result.notImplemented();
            result = null;
        }
    }

    private synchronized void start(@NonNull MethodCall call) {
        HashMap<String, Object> argumentsMap = (HashMap<String, Object>) call.arguments;

        // Mobile token
        String mobileToken = (String) argumentsMap.get("mobileToken");

        // PersonID
        String personId = (String) argumentsMap.get("personId");

        FaceAuthenticator.Builder mFaceAuthBuilder = new FaceAuthenticator.Builder(mobileToken);

        // Stage
        String stage = (String) argumentsMap.get("stage");
        if (stage != null) {
            mFaceAuthBuilder.setStage(CafStage.valueOf(stage));
        }

        // Filter
        String filter = (String) argumentsMap.get("filter");
        if (filter != null) {
            mFaceAuthBuilder.setFilter(Filter.valueOf(filter));
        }

        // Enable Screenshot
        Boolean enableScreenshot = (Boolean) argumentsMap.get("enableScreenshot");
        if (enableScreenshot != null) {
            mFaceAuthBuilder.setEnableScreenshots(enableScreenshot);
        }

        // FaceAuthenticator build
        mFaceAuthBuilder.build().authenticate(context, personId, new VerifyAuthenticationListener() {
            @Override
            public void onSuccess(FaceAuthenticatorResult faceAuthenticatorResult) {
                result.success(getSucessResponseMap(faceAuthenticatorResult));

            }

            @Override
            public void onError(FaceAuthenticatorResult faceAuthenticatorResult) {
                result.success(getFailureResponseMap(faceAuthenticatorResult));
            }

            @Override
            public void onCancel(FaceAuthenticatorResult faceAuthenticatorResult) {
                result.success(getClosedResponseMap());
            }

            @Override
            public void onLoading() {

            }

            @Override
            public void onLoaded() {

            }
        });
    }

    private HashMap<String, Object> getSucessResponseMap(FaceAuthenticatorResult result) {
        HashMap<String, Object> responseMap = new HashMap<>();
        responseMap.put("success", Boolean.TRUE);
        responseMap.put("signedResponse", result.getSignedResponse());

        return responseMap;
    }

    private HashMap<String, Object> getFailureResponseMap(FaceAuthenticatorResult result) {
        HashMap<String, Object> responseMap = new HashMap<>();
        responseMap.put("success", Boolean.FALSE);
        responseMap.put("errorMessage", result.getErrorMessage());
        return responseMap;
    }

    private HashMap<String, Object> getClosedResponseMap() {
        HashMap<String, Object> responseMap = new HashMap<>();
        responseMap.put("success", null);
        return responseMap;
    }

    @Override
    public synchronized boolean onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        return false;
    }

    @Override
    public synchronized void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
        this.context = flutterPluginBinding.getApplicationContext();
        this.channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "face_authenticator");
        this.channel.setMethodCallHandler(this);
    }

    @Override
    public synchronized void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
        this.channel.setMethodCallHandler(null);
        this.context = null;
    }

    @Override
    public synchronized void onAttachedToActivity(@NonNull ActivityPluginBinding binding) {
        this.activity = binding.getActivity();
        this.activityBinding = binding;
        this.activityBinding.addActivityResultListener(this);
    }

    @Override
    public synchronized void onDetachedFromActivityForConfigChanges() {

    }

    @Override
    public synchronized void onReattachedToActivityForConfigChanges(@NonNull ActivityPluginBinding binding) {
        this.activity = binding.getActivity();
    }

    @Override
    public synchronized void onDetachedFromActivity() {
        this.activity = null;
        this.activityBinding.removeActivityResultListener(this);
        this.activityBinding = null;
    }
}

File: android/src/main/kotlin/com/caf/face_authenticator/FaceAuthenticatorPlugin.kt
package com.caf.face_authenticator

import android.content.Context
import android.os.Looper
import input.CafStage
import input.FaceAuthenticator
import input.VerifyAuthenticationListener
import input.iproov.Filter
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import output.FaceAuthenticatorResult
import output.failure.NetworkReason
import output.failure.ServerReason

class FaceAuthenticatorPlugin: FlutterPlugin {

    private lateinit var eventChannel: EventChannel
    private lateinit var methodChannel: MethodChannel

    private var eventSink: EventChannel.EventSink? = null
    private var flutterPluginBinding: FlutterPlugin.FlutterPluginBinding? = null

    private val methodCallHandler = MethodChannel.MethodCallHandler { call, result ->
        val context: Context? = flutterPluginBinding?.applicationContext

        if (call.method == "start") {
            start(call)
        } else {
            result.notImplemented()
        }
    }

    private fun start(call: MethodCall) {
        val context: Context? = flutterPluginBinding?.applicationContext

        val argumentsMap = call.arguments as HashMap<*, *>

        // Mobile token
        val mobileToken = argumentsMap["mobileToken"] as String?
        // PersonID
        val personId = argumentsMap["personId"] as String?

        val mFaceAuthenticatorBuilder = FaceAuthenticator.Builder(mobileToken)

        // Stage
        val stage = argumentsMap["stage"] as String?
        if (stage != null) {
            mFaceAuthenticatorBuilder.setStage(CafStage.valueOf(stage))
        }

        // Filter
        val filter = argumentsMap["filter"] as String?
        if (filter != null) {
            mFaceAuthenticatorBuilder.setFilter(Filter.valueOf(filter))
        }

        // Enable Screenshot
        val enableScreenshot = argumentsMap["enableScreenshot"] as Boolean?
        if (enableScreenshot != null) {
            mFaceAuthenticatorBuilder.setEnableScreenshots(enableScreenshot)
        }

        // Enable SDK default loading screen
        val enableLoadingScreen = argumentsMap["enableLoadingScreen"] as Boolean?
        if (enableLoadingScreen != null) {
            mFaceAuthenticatorBuilder.setLoadingScreen(enableLoadingScreen)
        }

        //FaceAuth build
        mFaceAuthenticatorBuilder.build().authenticate(context, personId, object : VerifyAuthenticationListener {
            override fun onSuccess(faceAuthResult: FaceAuthenticatorResult) {
                android.os.Handler(Looper.getMainLooper()).post {
                    eventSink?.success(getSuccessResponseMap(faceAuthResult))
                    eventSink?.endOfStream()
                }
            }

            override fun onError(faceAuthResult: FaceAuthenticatorResult) {
                android.os.Handler(Looper.getMainLooper()).post {
                    if (faceAuthResult.sdkFailure != null) {
                        eventSink?.success(getErrorResponseMap(faceAuthResult))
                    } else {
                        eventSink?.success(getFailureResponseMap(faceAuthResult))
                    }
                    eventSink?.endOfStream()
                }
            }

            override fun onCancel(faceAuthResult: FaceAuthenticatorResult) {
                android.os.Handler(Looper.getMainLooper()).post {
                    eventSink?.success(getClosedResponseMap())
                    eventSink?.endOfStream()
                }
            }

            override fun onLoading() {
                android.os.Handler(Looper.getMainLooper()).post {
                    eventSink?.success(getConnectingResponseMap())
                }
            }

            override fun onLoaded() {
                android.os.Handler(Looper.getMainLooper()).post {
                    eventSink?.success(getConnectedResponseMap())
                }
            }

        })
    }

    private fun getSuccessResponseMap(result: FaceAuthenticatorResult): HashMap<String, Any> {
        val responseMap = HashMap<String, Any>()
        responseMap["event"] = "success"
        responseMap["signedResponse"] = result.signedResponse
        return responseMap
    }

    private fun getErrorResponseMap(result: FaceAuthenticatorResult): HashMap<String, Any> {
        val responseMap = HashMap<String, Any>()
        responseMap["event"] = "error"

        if (result.sdkFailure is ServerReason) {
            responseMap["errorType"] = "ServerReason"
            responseMap["errorMessage"] = result.sdkFailure.message
            responseMap["code"] = (result.sdkFailure as ServerReason).code
        } else if (result.sdkFailure is NetworkReason) {
            responseMap["errorType"] = "NetworkReason"
            responseMap["errorMessage"] = result.sdkFailure.message
        }
        return responseMap
    }

    private fun getFailureResponseMap(result: FaceAuthenticatorResult): HashMap<String, Any> {
        val responseMap = HashMap<String, Any>()
        responseMap["event"] = "failure"
        responseMap["errorMessage"] = result.errorMessage
        return responseMap
    }

    private fun getClosedResponseMap(): HashMap<String, Any> {
        val responseMap = HashMap<String, Any>()
        responseMap["event"] = "canceled"
        return responseMap
    }

    private fun getConnectingResponseMap(): HashMap<String, Any> {
        val responseMap = HashMap<String, Any>()
        responseMap["event"] = "connecting"
        return responseMap
    }

    private fun getConnectedResponseMap(): HashMap<String, Any> {
        val responseMap = HashMap<String, Any>()
        responseMap["event"] = "connected"
        return responseMap
    }

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        this.flutterPluginBinding = binding

        methodChannel = MethodChannel(flutterPluginBinding!!.binaryMessenger, "face_authenticator")
        methodChannel.setMethodCallHandler(methodCallHandler)

        eventChannel = EventChannel(flutterPluginBinding!!.binaryMessenger, "liveness_listener")
        eventChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                eventSink = events
            }

            override fun onCancel(arguments: Any?) {
                eventSink = null
            }
        })
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        this.flutterPluginBinding = null
    }
}